app/models/authenticated_session.rb

```diff
@@ -45,6 +45,16 @@ def save
     end
 
     false
+  rescue ActiveRecord::RecordInvalid
+    self.authentication_method = nil
+    if one_time_password.nil?
```

Comment:
I suppose you could extract lines 38-44 into a method called, say `verify_or_resend_otp`, and then this `save` method would become something like:
```ruby
def save
  return false if !valid? || !actionable?
  verify_or_resend_otp
rescue ActiveRecord::RecordInvalid
  # Retry in case of race condition
  self.authentication_method = nil
  verify_or_resend_otp
end
```

Reply:
That's a better  name than what I came up with on my own; so I'll do that.